MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.h

```diff
@@ -78,7 +87,30 @@ extern EFI_DRIVER_SUPPORTED_EFI_VERSION_PROTOCOL  gNvmExpressDriverSupportedEfiV
 //
 #define NVME_ASYNC_CCQ_SIZE  255
 
-#define NVME_MAX_QUEUES  3                              // Number of queues supported by the driver
+// MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+// Maximum number of queue pairs supported by the driver, including the admin queues.
+// Queue 0 - Admin
+// Queue 1 - Blocking I/O (BlockIo Protocol)
+// Queue 2 - Asynchronous I/O (BlockIo2 Protocol)
+#define NVME_MAX_QUEUES  3
+
+// Returns if the controller supports the BlockIo2 protocol.
+// The BlockIo2 protocol is only supported if the controller has more than 1 queue pair allocated
+#define NVME_SUPPORT_BLOCKIO2(ContollerPointer)  (((ContollerPointer)->NumberOfIoQueuePairs) > 1)
+// MU_CHANGE [END] - Request Number of Queues from Controller
+
+// MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+//
+// Returns the number of pages required for a submission/completion queue
+// The Indices are the same as above, 0 for admin, 1 for blocking I/O, 2 for async I/O.
+//
+#define NVME_SQ_SIZE_IN_PAGES(ControllerPointer, Index) \

```

Comment:
Are any of these fields controlled by the device? I see based on this that we allocate data?